lib/common/zstdmt_depthreadpool.c

```diff
@@ -0,0 +1,178 @@
+		ctx->nbJobsRemaining--;
+		if (!ctx->nbJobsRemaining) break;
+		ZSTDMT_checkStatus(pthread_cond_signal(&ctx->cond), "Cond signal");
+		ZSTDMT_checkStatus(pthread_mutex_unlock(&ctx->mutex), "Unlock ctx");
+	}
+	ZSTDMT_checkStatus(pthread_cond_signal(&ctx->cond), "Cond signal");
+	ZSTDMT_checkStatus(pthread_mutex_unlock(&ctx->mutex), "Unlock ctx");
+	ZSTDMT_checkStatus(pthread_mutex_unlock(&thread->mutex), "Unlock thread");
+
+	return NULL;
+}
+
+/***************************************
+* ZSTDMT_depThreadPool API
+***************************************/
+
+ZSTDMT_DepThreadPoolCtx* ZSTDMT_depThreadPool_createCtx(size_t maxNbJobs, size_t maxNbThreads)
+{
+	ZSTDMT_DepThreadPoolCtx* ctx = (ZSTDMT_DepThreadPoolCtx*)malloc(sizeof(ZSTDMT_DepThreadPoolCtx));
+	size_t i;
+
+	if (ctx == NULL) err_abort(-1, "Malloc ctx");
+
+	ZSTDMT_checkStatus(pthread_mutex_init(&ctx->mutex, NULL), "Mutex init ctx");
+	ZSTDMT_checkStatus(pthread_cond_init(&ctx->cond, NULL), "Cond init ctx");
+	ctx->nbJobs = 0;
+	ctx->nbJobsRemaining = maxNbJobs;
+	ctx->maxNbThreads = maxNbThreads;
+	ctx->threads = (ZSTDMT_DepThreadPoolThread**)malloc(sizeof(ZSTDMT_DepThreadPoolThread*) * ctx->maxNbThreads);
+	if (ctx->threads == NULL) err_abort(-1, "Malloc threads");
+	ctx->jobs = (ZSTDMT_DepThreadPoolJob**)malloc(sizeof(ZSTDMT_DepThreadPoolJob*) * ctx->nbJobsRemaining);
+	if (ctx->jobs == NULL) err_abort(-1, "Malloc jobs");
+	for (i = 0; i < ctx->maxNbThreads; ++i) {
+		ctx->threads[i] = (ZSTDMT_DepThreadPoolThread*)malloc(sizeof(ZSTDMT_DepThreadPoolThread));
+		if (ctx->threads[i] == NULL) err_abort(-1, "Malloc thread");
+		ctx->threads[i]->ctx = ctx;
+		ctx->threads[i]->job = NULL;
+		ZSTDMT_checkStatus(pthread_mutex_init(&ctx->threads[i]->mutex, NULL), "Mutex init thread");
+		ZSTDMT_checkStatus(pthread_create(&ctx->threads[i]->thread, NULL, ZSTDMT_threadRoutine, ctx->threads[i]), "Thread create");
+	}
+
+	return ctx;
+}
+
+void ZSTDMT_depThreadPool_destroyCtx(ZSTDMT_DepThreadPoolCtx* ctx)
+{
+	size_t i;
+	for (i = 0; i < ctx->maxNbThreads; ++i) pthread_join(ctx->threads[i]->thread, NULL);
+	for (i = 0; i < ctx->maxNbThreads; ++i) {
+		ZSTDMT_checkStatus(pthread_mutex_destroy(&ctx->threads[i]->mutex), "Mutex destroy thread");
+		free(ctx->threads[i]);
+	}
+	free(ctx->threads);
+	free(ctx->jobs);
+	ZSTDMT_checkStatus(pthread_mutex_destroy(&ctx->mutex), "Mutex destroy ctx");
+	ZSTDMT_checkStatus(pthread_cond_destroy(&ctx->cond), "Cond destroy ctx");
+	free(ctx);
+}
+
+size_t ZSTDMT_depThreadPool_addJob(ZSTDMT_DepThreadPoolCtx* ctx, ZSTDMT_depThreadPoolFn fn,
+	void* data, size_t nbDeps, size_t* depJobIds)
+{
+	ZSTDMT_DepThreadPoolJob* job = (ZSTDMT_DepThreadPoolJob*)malloc(sizeof(ZSTDMT_DepThreadPoolJob));
+	if (job == NULL) err_abort(-1, "Malloc job");
+	job->fn = fn;
+	job->data = data;
+	job->started = 0;
+	job->finished = 0;
+	job->nbDeps = nbDeps;
+	job->depJobIds = (size_t*)malloc(sizeof(size_t) * nbDeps);
+	if (job->depJobIds == NULL) err_abort(-1, "Malloc depJobIds");
```

Comment:
note : if using this branch `(job->depJobIds == NULL)`, there is a memory leak with regards to `ZSTDMT_DepThreadPoolJob* job` which has been already allocated.

Reply:
Hmm, should I free job here after the branch? Seems clumsy. What do you think is the best way to handle something like this?

Reply:
Sorry, I missed that `err_abort()` was actually aborting the process,
so forget my earlier comment, it's fine.

A higher level discussion could be : should the program recover from such an allocation error ? or give up ?
And I believe that "give up" is the right call here.